File: mas_agents/src/run-eval.ts
/**
 * BrowseComp-Plus multi-agent evaluation runner.
 *
 * Runs a single architecture against queries and saves JSON outputs
 * in the BrowseComp-Plus format under runs/<run_name>/.
 *
 * Usage:
 *   bun run src/run-eval.ts \
 *     --architecture single \
 *     --model gpt-4.1 \
 *     --index-path indexes/bm25/ \
 *     --limit 1
 */

import { parseArgs } from "util";
import { resolve, join } from "path";
import { existsSync, mkdirSync, readdirSync, readFileSync, writeFileSync } from "fs";
import { CopilotClient } from "@github/copilot-sdk";
import { BM25Bridge } from "./bm25-bridge.js";
import { CopilotSingleAgent } from "./single-agent.js";
import {
  CopilotIndependentAgent,
  CopilotCentralizedAgent,
  CopilotDecentralizedAgent,
  CopilotHybridAgent,
} from "./multi-agents.js";
import type { AgentOutput } from "./base-agent.js";

const ARCHITECTURES = [
  "single",
  "independent",
  "centralized",
  "decentralized",
  "hybrid",
] as const;

type Architecture = (typeof ARCHITECTURES)[number];

function createAgent(
  arch: Architecture,
  model: string,
  client: CopilotClient,
  retriever: BM25Bridge,
  numAgents: number
) {
  switch (arch) {
    case "single":
      return new CopilotSingleAgent(model, client, retriever);
    case "independent":
      return new CopilotIndependentAgent(model, client, retriever, numAgents);
    case "centralized":
      return new CopilotCentralizedAgent(model, client, retriever, numAgents, 5);
    case "decentralized":
      return new CopilotDecentralizedAgent(model, client, retriever, numAgents, 3);
    case "hybrid":
      return new CopilotHybridAgent(model, client, retriever, numAgents, 1, 1);
  }
}

async function main() {
  const { values } = parseArgs({
    args: Bun.argv.slice(2),
    options: {
      architecture: { type: "string", short: "a" },
      model: { type: "string", short: "m", default: "gpt-4.1" },
      query: { type: "string", short: "q", default: "topics-qrels/queries.tsv" },
      "index-path": { type: "string", short: "i", default: "indexes/bm25/" },
      "output-dir": { type: "string", short: "o" },
      k: { type: "string", default: "5" },
      "snippet-max-tokens": { type: "string", default: "512" },
      "num-agents": { type: "string", default: "3" },
      limit: { type: "string", default: "0" },
    },
    strict: true,
  });

  const arch = values.architecture as Architecture;
  if (!arch || !ARCHITECTURES.includes(arch)) {
    console.error(
      `Error: --architecture must be one of: ${ARCHITECTURES.join(", ")}`
    );
    process.exit(1);
  }

  const model = values.model!;
  const queryPath = values.query!;
  const indexPath = values["index-path"]!;
  const k = parseInt(values.k!);
  const snippetMaxTokens = parseInt(values["snippet-max-tokens"]!);
  const numAgents = parseInt(values["num-agents"]!);
  const limit = parseInt(values.limit!);

  // Resolve output directory
  const modelSlug = model.replace(/\//g, "_").replace(/-/g, "_");
  const outputDir = values["output-dir"]
    ? resolve(values["output-dir"])
    : resolve(`runs/bm25/${arch}_${modelSlug}`);

  mkdirSync(outputDir, { recursive: true });

  // ── Load queries ────────────────────────────────────────────────────────
  if (!existsSync(queryPath)) {
    console.error(`Error: Query file not found: ${queryPath}`);
    process.exit(1);
  }

  const queryContent = readFileSync(queryPath, "utf-8");
  let queries: Array<[string, string]> = queryContent
    .trim()
    .split("\n")
    .map((line) => {
      const [id, ...rest] = line.split("\t");
      return [id.trim(), rest.join("\t").trim()] as [string, string];
    })
    .filter(([id, text]) => id && text);

  console.log(`Loaded ${queries.length} queries from ${queryPath}`);

  if (limit > 0) {
    queries = queries.slice(0, limit);
    console.log(`  Limited to ${limit} queries`);
  }

  // ── Resume: skip already-processed queries ──────────────────────────────
  const processedIds = new Set<string>();
  if (existsSync(outputDir)) {
    for (const file of readdirSync(outputDir)) {
      if (file.startsWith("run_") && file.endsWith(".json")) {
        try {
          const data = JSON.parse(
            readFileSync(join(outputDir, file), "utf-8")
          );
          if (data.query_id) processedIds.add(String(data.query_id));
        } catch {
          // skip
        }
      }
    }
  }

  const remaining = queries.filter(([qid]) => !processedIds.has(qid));
  console.log(
    `Processing ${remaining.length} remaining queries (skipping ${processedIds.size} already done)`
  );

  if (remaining.length === 0) {
    console.log("All queries already processed. Done.");
    return;
  }

  // ── Initialize BM25 bridge ──────────────────────────────────────────────
  console.log(`Initializing BM25 retriever from ${indexPath}...`);
  const retriever = new BM25Bridge({
    indexPath,
    k,
    snippetMaxTokens,
  });
  await retriever.start();

  // ── Initialize Copilot client ───────────────────────────────────────────
  console.log("Starting Copilot SDK client...");
  const client = new CopilotClient();

  // ── Create agent ────────────────────────────────────────────────────────
  const agent = createAgent(arch, model, client, retriever, numAgents);
  console.log(`Agent: ${agent.constructor.name} | Model: ${model}`);
  console.log(`Output: ${outputDir}`);
  console.log("=".repeat(60));

  // ── Process queries ─────────────────────────────────────────────────────
  let completed = 0;
  let failed = 0;

  for (let idx = 0; idx < remaining.length; idx++) {
    const [qid, qtext] = remaining[idx];
    const progress = `[${idx + 1}/${remaining.length}]`;

    try {
      const result: AgentOutput = await agent.runQuery(qid, qtext);

      // Save result
      const ts = new Date().toISOString().replace(/[-:]/g, "").replace(/\..+/, "Z");
      const filename = join(outputDir, `run_${ts}_${qid}.json`);
      writeFileSync(filename, JSON.stringify(result, null, 2), "utf-8");

      const status = result.status;
      const searchCount = result.tool_call_counts?.search ?? 0;
      console.log(`  ${progress} [${qid}] status=${status} | searches=${searchCount}`);

      if (status === "completed") completed++;
      else failed++;
    } catch (exc) {
      console.error(`  ${progress} [${qid}] FAILED: ${exc}`);
      failed++;

      const ts = new Date().toISOString().replace(/[-:]/g, "").replace(/\..+/, "Z");
      const filename = join(outputDir, `run_${ts}_${qid}.json`);
      writeFileSync(
        filename,
        JSON.stringify(
          {
            query_id: qid,
            tool_call_counts: {},
            status: `error: ${exc}`,
            retrieved_docids: [],
            result: [{ type: "output_text", output: "" }],
          },
          null,
          2
        ),
        "utf-8"
      );
    }
  }

  // ── Summary ─────────────────────────────────────────────────────────────
  const total = completed + failed;
  console.log("\n" + "=".repeat(60));
  console.log(`Done! ${completed}/${total} completed, ${failed}/${total} failed`);
  console.log(`Results saved to: ${outputDir}`);

  // Cleanup
  await retriever.stop();
  await client.stop();
  process.exit(0);
}

main().catch((e) => {
  console.error("Fatal error:", e);
  process.exit(1);
});

File: mas_agents/run_eval.py
"""
BrowseComp-Plus multi-agent evaluation runner.

Runs a single architecture against all 100 queries and saves JSON outputs
in the BrowseComp-Plus format under runs/<run_name>/.

Usage:
    uv run python mas_agents/run_eval.py \
        --architecture single \
        --model gpt-5-mini \
        --index-path indexes/bm25/ \
        --output-dir runs/bm25/single_gpt5mini
"""

import argparse
import asyncio
import csv
import json
import os
import sys
from datetime import datetime
from pathlib import Path

from tqdm import tqdm


async def main():
    parser = argparse.ArgumentParser(
        description="Run BrowseComp-Plus evaluation with MAS architectures"
    )
    parser.add_argument(
        "--architecture",
        choices=["single", "independent", "centralized", "decentralized", "hybrid"],
        required=True,
        help="Agent architecture to evaluate",
    )
    parser.add_argument(
        "--model",
        default="gpt-5-mini",
        help="Copilot model name (default: %(default)s)",
    )
    parser.add_argument(
        "--query",
        default="topics-qrels/queries.tsv",
        help="Path to queries TSV file (default: %(default)s)",
    )
    parser.add_argument(
        "--index-path",
        default="indexes/bm25/",
        help="Path to BM25 Lucene index (default: %(default)s)",
    )
    parser.add_argument(
        "--output-dir",
        default=None,
        help="Directory to save results (default: runs/bm25/<architecture>_<model>)",
    )
    parser.add_argument(
        "--k",
        type=int,
        default=5,
        help="Number of search results per query (default: %(default)s)",
    )
    parser.add_argument(
        "--snippet-max-tokens",
        type=int,
        default=512,
        help="Max tokens per document snippet (default: %(default)s)",
    )
    parser.add_argument(
        "--num-agents",
        type=int,
        default=3,
        help="Number of sub-agents for MAS architectures (default: %(default)s)",
    )
    parser.add_argument(
        "--limit",
        type=int,
        default=0,
        help="Limit number of queries to process (0=all, default: %(default)s)",
    )

    args = parser.parse_args()

    # Resolve output directory
    if args.output_dir is None:
        model_slug = args.model.replace("/", "_").replace("-", "_")
        args.output_dir = f"runs/bm25/{args.architecture}_{model_slug}"

    out_dir = Path(args.output_dir).resolve()
    out_dir.mkdir(parents=True, exist_ok=True)

    # ── Load queries ──────────────────────────────────────────────────────────
    query_path = Path(args.query)
    if not query_path.is_file():
        print(f"Error: Query file not found: {query_path}")
        sys.exit(1)

    queries: list[tuple[str, str]] = []
    with query_path.open(newline="", encoding="utf-8") as f:
        reader = csv.reader(f, delimiter="\t")
        for row in reader:
            if len(row) >= 2:
                queries.append((row[0].strip(), row[1].strip()))

    print(f"Loaded {len(queries)} queries from {query_path}")

    if args.limit > 0:
        queries = queries[: args.limit]
        print(f"  Limited to {args.limit} queries")

    # ── Resume: skip already-processed queries ────────────────────────────────
    processed_ids: set[str] = set()
    if out_dir.exists():
        for json_path in out_dir.glob("run_*.json"):
            try:
                with json_path.open("r", encoding="utf-8") as jf:
                    meta = json.load(jf)
                    qid = meta.get("query_id")
                    if qid:
                        processed_ids.add(str(qid))
            except Exception:
                continue

    remaining = [(qid, qtext) for qid, qtext in queries if qid not in processed_ids]
    print(f"Processing {len(remaining)} remaining queries (skipping {len(processed_ids)} already done)")

    if not remaining:
        print("All queries already processed. Done.")
        return

    # ── Initialize retriever ──────────────────────────────────────────────────
    from .base_agent import RetrieverTool

    print(f"Initializing BM25 retriever from {args.index_path}...")
    retriever = RetrieverTool(
        index_path=args.index_path,
        k=args.k,
        snippet_max_tokens=args.snippet_max_tokens,
    )

    # ── Initialize Copilot client ─────────────────────────────────────────────
    from .llm_client import get_copilot_client

    print(f"Starting Copilot SDK client...")
    client = await get_copilot_client()

    # ── Create agent ──────────────────────────────────────────────────────────
    agent = _create_agent(args.architecture, args.model, client, retriever, args.num_agents)
    print(f"Agent: {agent.__class__.__name__} | Model: {args.model}")
    print(f"Output: {out_dir}")
    print("=" * 60)

    # ── Process queries ───────────────────────────────────────────────────────
    completed = 0
    failed = 0

    for qid, qtext in tqdm(remaining, desc="Queries", unit="query"):
        try:
            result = await agent.run_query(qid, qtext)

            # Save result
            ts = datetime.utcnow().strftime("%Y%m%dT%H%M%SZ")
            filename = out_dir / f"run_{ts}_{qid}.json"
            with filename.open("w", encoding="utf-8") as f:
                json.dump(result, f, indent=2, default=str)

            status = result.get("status", "unknown")
            search_count = result.get("tool_call_counts", {}).get("search", 0)
            print(f"  [{qid}] status={status} | searches={search_count}")

            if status == "completed":
                completed += 1
            else:
                failed += 1

        except Exception as exc:
            print(f"  [{qid}] FAILED: {exc}")
            failed += 1

            # Save failure record
            ts = datetime.utcnow().strftime("%Y%m%dT%H%M%SZ")
            filename = out_dir / f"run_{ts}_{qid}.json"
            with filename.open("w", encoding="utf-8") as f:
                json.dump({
                    "query_id": qid,
                    "tool_call_counts": {},
                    "status": f"error: {exc}",
                    "retrieved_docids": [],
                    "result": [{"type": "output_text", "output": ""}],
                }, f, indent=2)

    # ── Summary ───────────────────────────────────────────────────────────────
    total = completed + failed
    print("\n" + "=" * 60)
    print(f"Done! {completed}/{total} completed, {failed}/{total} failed")
    print(f"Results saved to: {out_dir}")

    # Stop Copilot client
    try:
        await client.stop()
    except Exception:
        pass


def _create_agent(architecture: str, model: str, client, retriever, num_agents: int):
    """Factory function to create the appropriate agent."""
    from .single_agent import CopilotSingleAgent
    from .multi_agents import (
        CopilotIndependentAgent,
        CopilotCentralizedAgent,
        CopilotDecentralizedAgent,
        CopilotHybridAgent,
    )

    if architecture == "single":
        return CopilotSingleAgent(model, client, retriever)
    elif architecture == "independent":
        return CopilotIndependentAgent(model, client, retriever, num_agents=num_agents)
    elif architecture == "centralized":
        return CopilotCentralizedAgent(model, client, retriever, num_agents=num_agents, rounds=5)
    elif architecture == "decentralized":
        return CopilotDecentralizedAgent(model, client, retriever, num_agents=num_agents, rounds=3)
    elif architecture == "hybrid":
        return CopilotHybridAgent(model, client, retriever, num_agents=num_agents, rounds=1, peer_rounds=1)
    else:
        raise ValueError(f"Unknown architecture: {architecture}")


if __name__ == "__main__":
    asyncio.run(main())
